src/events/message/commands/fun/rankup.ts

```diff
@@ -0,0 +1,184 @@
+import { z } from "zod";
+import fetch from "node-fetch";
+import { Message, CommandArg, word } from "../types";
+
+const USAGE: string =
+  "Usage: `?rankup [<username> --target=(username|rank) --language=(py|js|...) --mode={least|each|spread} --limit=(1-8[kyu])]`";
+
+/*
+`?rankup` calculates the number of katas of each rank required for a user to rank up.
+This command is just for fun, however should increase competition between users,
+and (hopefully) increase motivation.
+
+If no username is passed, server nickname of user is taken instead.
+
+Options:
+  <username>   The username to check. Defaults to users discord server nickname
+  --target=    The rank to reach, or a username of a user to overtake. Defaults to users next rank
+  --language=  The language to inspect. Defaults to "overall"
+  --mode=      Calculation mode.
+                  Least favours 1kyus for least overall katas.
+                  Each finds number of katas per rank individually
+                  Spread (default) prioritises high katas, but also distributes points to lower ones
+  --limit=     The max rank to use in calculations ([1-8]kyu)
+*/
+
+enum Mode {
+  Least = "least",
+  Each = "each",
+  Spread = "spread", // default
+}
+const DEFAULTMODE: Mode = Mode.Spread;
+const SPREADWEIGHT: number = 0.6; // used for Spread mode
+
+// Convert synonyms into the names used by CW API
+const LANGSYNONYMS: { [lang: string]: string } = {
+  py: "python",
+  js: "javascript",
+  hs: "haskell",
+  brainfuck: "bf",
+  cs: "coffeescript",
+  "c#": "csharp",
+  ts: "typescript",
+  "c++": "cpp",
+  "f#": "fsharp",
+  objectivec: "objc",
+  "obj-c": "objc",
+  "objective-c": "objc",
+  coldfusionmarkuplanguage: "cfml",
+  coldfusion: "cfml",
+};
+
+// https://docs.codewars.com/gamification/ranks
+// 3-8 dan added speculatively, based on trend
+const RANKTHRESHOLDS: { [rank: string]: number } = {
```

Comment:
The formula should be `floor(exp(rank)*12 - 12)` where rank is `0..=15` for 8 kyu to 8 dan.